Docs: add some introductory material, vector bases and SEM descriptions (#602)
CliMA/ClimaCore.jl: docs/src/api.md

Geometry.float_type
```

Points represent _locations_ in space (not to be confused with _positions_,

> _position vectors_ or _location vectors_ in geometry). Vectors, on the other hand,

Reviewer: I still don't understand the distinction between a coordinate and a position.

Author: This comment was here only for the people with a geometry background (like me) that might be confused because in geometry the terminology _position_ can be used for [_position vector_ or _location vector_](https://en.wikipedia.org/wiki/Position_(geometry)#:~:text=The%20term%20%22position%20vector%22%20is,affine%20spaces%20of%20any%20dimension.), i.e., a Euclidean vector that represents the position of a point P in space in relation to an arbitrary reference origin O. But if not useful for this community, I can certainly remove it.

Reviewer: Having thought about this a bit more, I think the more useful distinction is that a "point' describes a position via a particular set of coordinates, but that that the coordinates need not be Cartesian.

Author: Yes. So do we prefer to rephrase the sentence above or is it good as is?

Reviewer: I would prefer to rephrase it if possible, as I think the current wording is confusing.

Author: Sure, what do you have in mind?

Author: @simonbyrne , I applied your other suggestions. For this point in particular I am not sure what you had in mind. As I said, we can just remove the part in parentheses. I don't mind.

> Points represent _locations_ in space (not to be confused with _positions_,
> _position vectors_ or _location vectors_ in geometry). Vectors, on the other hand,
> represent displacements in space. An analogy with time works well: times,
> (also called instants or datetimes) are _locations_ in time. While, durations
> are displacements in time.

Reviewer: ```suggestion
Points represent _locations_ in space, specified by a coordinates in a given coordinate system (Cartesian, spherical, etc), whereas vectors, on the other hand, represent displacements in space. 
An analogy with time works well: times (also called instants or datetimes) are _locations_ in time, while, durations are displacements in time.
```

> **Note 1**: In ClimaCore.jl, we use angles (and, therefore, trigonometric functions:
> `cosd`, `sind`, `acosd`, `asind`, `tand`,...) in degrees, not in radians.

Reviewer: ```suggestion
**Note 1**: Latitude and longitude are specified via angles (and, therefore, trigonometric functions:
`cosd`, `sind`, `acosd`, `asind`, `tand`,...) in degrees, not in radians.
```

```

## Spaces
A `Space` represents a discretized function space over some domain.
Currently two main discretizations are supported: Spectral Element Discretization
(both Continuous Galerkin and Discontinuous Galerkin types) and a staggered
Finite Difference Discretization. Combination of these two in the horizontal/vertical
directions, respectively, is what we call a _hybrid_ space.

Sketch of a 2DX hybrid discretization:

![3D hybrid discretization in a Cartesian domain](DiscretizationSketch.png)

```@docs
Spaces
```
### Finite Difference Spaces
ClimaCore.jl supports staggered Finite Difference discretizations. Finite Differences
discretize an interval domain by approximating the function by a value at either
the center of each element (also referred to as _cell_) (`CenterFiniteDifferenceSpace`),
or the interfaces (faces in 3D, edges in 2D or points in 1D) between elements
(`FaceFiniteDifferenceSpace`).

> Users can construct either the center or face space from the mesh, then construct
> the dual space from the original one (this is to avoid allocating additional memory).

Reviewer: ```suggestion
Users should construct either the center or face space from the mesh, then construct
the other space from the original one: this internally reuses the same datastructures, and avoids allocating additional memory.
```
